File: algorithm/课件/11/lc1143 最长公共子序列.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n = len(text1)
        m = len(text2)
        text1 = " " + text1
        text2 = " " + text2
        f = [[0] * (m + 1) for i in range(n + 1)] # 0~n 0~m
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                if text1[i] == text2[j]:
                    f[i][j] = f[i-1][j-1] + 1
                else:
                    f[i][j] = max(f[i-1][j], f[i][j-1])
        return f[n][m]

File: algorithm/课件/11/lc1143 最长公共子序列.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int n = text1.length();
        int m = text2.length();
        text1 = " " + text1;
        text2 = " " + text2;
        int[][] f = new int[n + 1][m + 1];
        for (int i = 0; i <= n; i++) f[i][0] = 0;
        for (int j = 0; j <= m; j++) f[0][m] = 0;
        for (int i = 1; i <= n; i++)
            for (int j =1; j <= m; j++)
                if (text1.charAt(i) == text2.charAt(j))
                    f[i][j] = f[i - 1][j - 1] + 1;
                else
                    f[i][j] = Math.max(f[i - 1][j], f[i][j - 1]);
        return f[n][m];
    }
}